src/Stratis.Bitcoin/Consensus/ConsensusManger.cs

```diff
@@ -74,13 +106,242 @@ public class ConsensusManager
+            }
+
+            this.logger.LogTrace("(-)");
+        }
+
+        /// <summary>
+        /// A callback that is triggered when a block was downloaded.
+        /// </summary>
+        private void ProcessDownloadedBlock(BlockPair blockPair)
+        {
+            this.logger.LogTrace("({0}:'{1}')", nameof(blockPair), blockPair);
+
+            bool partialValidationRequired = false;
+
+            lock (this.treeLock)
+            {
+                partialValidationRequired = this.chainedHeaderTree.BlockDataDownloaded(blockPair.ChainedHeader, blockPair.Block);
+            }
+
+            if (partialValidationRequired)
+            {
+                this.chainedHeaderValidator.StartPartialValidation(blockPair, this.OnPartialValidationCompletedCallback);
+            }
+
+            this.logger.LogTrace("(-)");
+        }
+
+        private void OnPartialValidationCompletedCallback(BlockPair blockPair, bool success)
+        {
+            this.logger.LogTrace("({0}:'{1}',{2}.{3}:{4})", nameof(blockPair), blockPair, nameof(success), success);
+
+            // TODO
 
             this.logger.LogTrace("(-)");
         }
+
+        private void DownloadBlocks(List<uint256> blockHashes, ProcessDownloadedBlockDelegate processesBlock)
+        {
+            this.logger.LogTrace("({0}.{1}:{2})", nameof(blockHashes), nameof(blockHashes.Count), blockHashes.Count);
+
+            BlockDownloadRequest request = new BlockDownloadRequest
+            {
+                BlocksToDownload = blockHashes,
+                DownloadedBlockDelegate = processesBlock
+            };
+
+            lock (this.treeLock)
+            {
+                this.toDownloadQueue.Enqueue(request);
+                this.ProcessDownloadQueueLocked();
+            }
+
+            this.logger.LogTrace("(-)");
+        }
+
+        /// <summary>
+        /// A method that to requests blocks, any block that is found is sent to the callback <see cref="processesBlock"/>.
+        /// If the block is not part of the tree (on a known the chain) the a <c>null</c> value will be return in the callback.
+        /// If the block is available in memory or on disk it will be immediately return the block on the callback.
+        /// The block hashes must be in a consecutive order.
+        /// </summary>
+        /// <param name="blockHashes">The block hashes to download, the block hashes must be in a consecutive order.</param>
+        /// <param name="processesBlock">The callback that will be called for each downloaded block.</param>
+        /// <returns></returns>
+        public async Task GetOrDownloadBlocksAsync(List<uint256> blockHashes, ProcessDownloadedBlockDelegate processesBlock)
+        {
+            this.logger.LogTrace("({0}.{1}:{2})", nameof(blockHashes), nameof(blockHashes.Count), blockHashes.Count);
+
+            // TODO: verify the blockHashes are consecutive
+
+            List<uint256> downloadJob = new List<uint256>();
```

Comment:
Some logic is missing in this method:
```
hashesToDownload.Add(hash) //Check consequtiveness here. If for some reason newly added pair.ChainedHeader.Prev != DownloadBlocks.Last - Call DownloadBlocks on the list we have and recreate the list
							// So we always have only consecutive headers in the list.
```

So the point here is if we are asked for hashes

a1=a2=a3=a4=a5=a9=a10=a12=a13=b2=b3=b4=b5

we should generate 3 download jobs:
1) a1=a2=a3=a4=a5
2) a9=a10=a12=a13
3) b2=b3=b4=b5

because block puller requires download job headers to be always consecutive.

Reply:
yes I missed that one thanks
Fixed